### PR TITLE
Allow individual upvars to be inferred as by-value or by-ref in non-escaping closures

### PR DESCRIPTION
**Pull request by nikomatsakis**
r? @eddyb

### CODE REVIEW

File: src/librustc_trans/trans/closure.rs [hunk 1]
```diff
@@ -100,8 +100,7 @@ fn load_unboxed_closure_environment<'blk, 'tcx>(
 #[derive(PartialEq)]
 pub enum ClosureKind<'tcx> {
```
Comment (eddyb) [inline src/librustc_trans/trans/closure.rs]: `'tcx` is not used here anymore.

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -304,7 +304,7 @@ impl<'a, 'tcx> mc::Typer<'tcx> for FnCtxt<'a, 'tcx> {
     }
     fn type_moves_by_default(&self, span: Span, ty: Ty<'tcx>) -> bool {
         let ty = self.infcx().resolve_type_vars_if_possible(&ty);
-        traits::type_known_to_meet_builtin_bound(self.infcx(), self, ty, ty::BoundCopy, span)
+        !traits::type_known_to_meet_builtin_bound(self.infcx(), self, ty, ty::BoundCopy, span)
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Uhm... what? How did this remain unnoticed until now?

File: src/librustc/middle/ty.rs [hunk 3]
```diff
@@ -1240,60 +1237,31 @@ pub enum BorrowKind {
...
+    ByValue,
+
+    /// Upvar is captured by reference.
+    ByRef(UpvarBorrow),
+}
+
 #[derive(PartialEq, Clone, RustcEncodable, RustcDecodable, Show, Copy)]
 pub struct UpvarBorrow {
+    /// The kind of borrow: by-ref upvars have access to shared
+    /// immutable borrows, which are not part of the normal language
```
Comment (eddyb) [inline src/librustc/middle/ty.rs]: "Shared immutable borrows" sounds like `&T` - did you mean to refer to `&uniq T` instead?

File: src/librustc_trans/trans/closure.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_trans/trans/closure.rs]: On a second note, `ClosureEnv` can now become `Closure(&'a [ty::Freevar]) | NotClosure`, which seems more idiomatic.

**Comment by eddyb** [comment]
r=me with nits fixed. There seems to be partial overlap with #21598, but nothing too complicated.

**Comment by aturon** [comment]
\o/

File: src/librustc_typeck/check/mod.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: This code path isn't used except for upvar inference.

File: src/librustc/middle/ty.rs (hunk 3 above)
Comment (nikomatsakis) [inline src/librustc/middle/ty.rs]: Yes, thanks.

**Comment by eddyb** [comment]
@nikomatsakis did you mean to include @kvark's commit into this?

**Comment by eddyb** [comment]
Oh it's in master already. @bors r- until that's fixed.

**Comment by nikomatsakis** [comment]
@eddyb corrected

**Comment by nikomatsakis** [comment]
Grr, this test failure is very strange! Certainly I didn't see it locally.

**Pull request merged**